lib/XML/LibXML/Dom.pm

```diff
@@ -0,0 +1,281 @@
+use v6.c;
+
+use XML::LibXML::CStructs :types;
+
+class XML::LibXML::xmlNs is xmlNs {
+
+	method setNext(xmlNs $n) {
+		$.next = $n;
+	}
+
+}
+
+package XML::LibXML::Dom {
+
+	use NativeCall;
+
+	use XML::LibXML::Enums;
+	use XML::LibXML::Subs;
+
+	sub _domAddNsChain(xmlNsPtr $c, xmlNsPtr $ns) {
+		return $ns unless $c =:= xmlNsPtr;
+
+		my $i = $c;
+		while ($i !=:= xmlNsPtr && $i !=:= $ns) {
+			$i = nativecast(xmlNs, $i).next;
+			if ($i =:= xmlNsPtr) {
+				nativecast(xmlNs, $ns).setNext($c);
+				return $ns;
+			}
+		}
+
+		return $c;
+	}
+
+	sub _domReconcileNsAttr(xmlAttrPtr $a, xmlNsPtr $unused) {
+		return unless $a.defined;
+
+		my $attr = nativecast(xmlAttr, $a);
+
+		my xmlNodePtr $tree = $attr.parent;
+		my xmlNode $tree_o = nativecast(xmlNode, $tree);
+
+		return if ! $tree.defined;
+
+		if ($attr.ns.defined) {
+			my xmlNsPtr $ns;
+
+			if ($attr.ns.name.defined && $attr.ns.name eq "xml") {
+				$ns = xmlSearchNsByHref($tree.doc, $tree, XML_XML_NAMESPACE);
+				$attr.setNs($ns);
+				return
+			} else {
+				$ns = xmlSearchNs($tree_o.doc, $tree_o.parent, $attr.ns.name);
+			}
+
+			my $nso = nativecast(xmlNs, $ns);
+			if [&&](
```

Comment:
why the fancy reduce op here? and old-school `if A && B && C` will do, and the &&'s look nicer. (you spot the begging of the block more easily)